cache.go

```diff
@@ -143,6 +143,38 @@ func (c *cache) get(k string) (interface{}, bool) {
 	return item.Object, true
 }
 
+func (c *cache) getWithExpiration(k string) (interface{}, time.Time, bool) {
+	item, found := c.items[k]
+	if !found {
+		return nil, time.Time{}, false
+	}
+	// "Inlining" of Expired
+	if item.Expiration > 0 {
+		if time.Now().UnixNano() > item.Expiration {
+			return nil, time.Time{}, false
+		}
+	}
+	return item.Object, time.Unix(0, item.Expiration), true
+}
+
+// SetTransaction given a values current state and expiration and whether it was found in the cache
+// atomically update and return the value and the new expiration time
+type SetTransaction func(interface{}, time.Time, bool) (interface{}, time.Duration)
+
+// GetAndSet allows retrieval of a cache key and setting the value based on a setFn in a synchronized atomic manner
+func (c *cache) GetAndSet(k string, setFn SetTransaction) (interface{}, bool) {
+	c.mu.RLock()
+	v, expiration, found := c.getWithExpiration(k)
+
+	newV, duration := setFn(v, expiration, found)
+
+	c.set(k, newV, duration)
+
+	c.mu.RUnlock()
```

Comment:
I think it should be done in a `defer` call just after locking (as there is a user defined function call before, a panic in it would break the mutex and leave it locked).